homeassistant/components/smartthings/config_flow.py

```diff
@@ -0,0 +1,186 @@
+"""Config flow to configure SmartThings."""
+import logging
+
+from aiohttp.client_exceptions import ClientResponseError
+import voluptuous as vol
+
+from homeassistant import config_entries
+from homeassistant.const import CONF_ACCESS_TOKEN
+from homeassistant.helpers.aiohttp_client import async_get_clientsession
+
+from .const import (
+    CONF_APP_ID, CONF_INSTALLED_APP_ID, CONF_LOCATION_ID, DOMAIN,
+    VAL_UID_MATCHER)
+from .smartapp import (
+    create_app, find_app, setup_smartapp, setup_smartapp_endpoint, update_app)
+
+_LOGGER = logging.getLogger(__name__)
+
+
+@config_entries.HANDLERS.register(DOMAIN)
+class SmartThingsFlowHandler(config_entries.ConfigFlow):
+    """
+    Handle configuration of SmartThings integrations.
+
+    Any number of integrations are supported.  The high level flow follows:
+    1) Flow initiated
+        a) User initiates through the UI
+        b) Re-configuration of a failed entry setup
+    2) Enter access token
+        a) Check not already setup
+        b) Validate format
+        c) Setup SmartApp
+    3) Wait for Installation
+        a) Check user installed into one or more locations
+        b) Config entries setup for all installations
+    """
+
+    VERSION = 1
+    CONNECTION_CLASS = config_entries.CONN_CLASS_CLOUD_PUSH
+
+    def __init__(self):
+        """Create a new instance of the flow handler."""
+        self.access_token = None
+        self.app_id = None
+        self.api = None
+
+    async def async_step_import(self, user_input=None):
+        """Occurs when a previously entry setup fails and is re-initiated."""
+        return await self.async_step_user(user_input)
+
+    async def async_step_user(self, user_input=None):
+        """Get access token and validate it."""
+        from pysmartthings import SmartThings
+
+        errors = {}
+        if not self.hass.config.api.base_url.lower().startswith('https://'):
+            errors['base'] = "base_url_not_https"
+            return self._show_step_user(errors)
+
+        if user_input is None or CONF_ACCESS_TOKEN not in user_input:
```

Comment:
That condition occurs when we receive a `install` lifecycle event over the webhook, which is when the user "installs" the SmartApp via the SmartThings mobile app.  We don't get the personal access token in the event data.  There are 2 scenarios for when this happens:
1.  While the user is within another config flow (on the waiting screen).  We discard (abort) this flow because the user config flow will create the entry.
2. After the user has completed the config flow and they install the SmartApp into another location.  We can grab the access token from another config entry with the same parent `app_id` as the token would be the same.

The check for the first scenario could be moved to the `smartapp_install` event handler to make this less confusing.

Reply:
I moved the logic to `smartapp_install` which simplified this greatly.